### PR TITLE
[SYCL] Add test for classes implicitly converted from items in parallel_for

### PR DESCRIPTION
**Pull request by aobolensk**

### CODE REVIEW

File: SYCL/Basic/parallel_for_user_types.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %t.out
```
Comment (vladimirlaz) [inline SYCL/Basic/parallel_for_user_types.cpp]:
```suggestion
// RUN: %HOST_RUN_PLACEHOLDER %t.out
// RUN: %CPU_RUN_PLACEHOLDER %t.out
// RUN: %GPU_RUN_PLACEHOLDER %t.out
// RUN: %ACC_RUN_PLACEHOLDER %t.out
```
We should run on all possible devices separately to verify all possible taret devices

File: SYCL/Basic/parallel_for_user_types.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  // Check user defined sycl::nd_item wrapper
+  q.submit([&](sycl::handler &h) {
+    auto buf_acc = data_buf.get_access<sycl::access::mode::read_write>(h);
+    h.parallel_for(sycl::nd_range<1>{sz, 2},
+                   [=](nd_item_wrapper<1> item) { buf_acc[item.get()] += 1; });
+  });
+
+  for (int i = 0; i < sz; ++i) {
+    assert(data[i] == i + 2);
```
Comment (vladimirlaz) [inline SYCL/Basic/parallel_for_user_types.cpp]: please do NOT use assert to verify functional correctness. Exit code can be used instead.

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: SYCL/Basic/parallel_for_user_types.cpp (hunk 2 above)
Comment (aobolensk) [reply SYCL/Basic/parallel_for_user_types.cpp]: Done

**Comment by vladimirlaz** [comment]
@aobolensk, could you please fix failures in the added test: see http://pre-commit.llvm.intel.com:8013/file_review?path_name=/llvm_test_suite_associate/PR5118/dc0e75b09f

File: SYCL/Basic/parallel_for_user_types.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+  // Check user defined sycl::nd_item wrapper
+  q.submit([&](sycl::handler &h) {
+    auto buf_acc = data_buf.get_access<sycl::access::mode::read_write>(h);
+    h.parallel_for(sycl::nd_range<1>{sz, 2},
+                   [=](nd_item_wrapper<1> item) { buf_acc[item.get()] += 1; });
+  });
+  q.wait();
+
+  for (int i = 0; i < sz; ++i) {
```
Comment (vladimirlaz) [inline SYCL/Basic/parallel_for_user_types.cpp]: The SYCL specification guarantees that the data from the device can be obtained on the host using a host accessor or after the buffer destruction (see https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_synchronization_in_the_sycl_application)

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

**Review by vladimirlaz — APPROVED** [review]

**Comment by vladimirlaz** [comment]
The tests below failed due to unrelated change.
The PR can be committed after intel/llvm#5118 is submitted and nightly build is ready. The tests have be restarted first.
```
[2021-12-13T08:32:59.136Z] Failed Tests (7):
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_kernels_win.cpp
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_multiple_tus_one_ndebug_win.cpp
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_multiple_tus_win.cpp
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_one_kernel_win.cpp
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_simultaneous_kernels_win.cpp
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_simultaneously_multiple_tus.cpp
[2021-12-13T08:32:59.136Z]   SYCL :: Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp
```

**Pull request merged**